hookman/hooks.py

```diff
@@ -0,0 +1,142 @@
+        If an error is found, a TypeError exception will be raised
+        """
+        hook_args = inspect.getfullargspec(hook)
+
+        if not hook_args.args:
+            raise TypeError("It's not possible to create a hook without argument")
+
+        annotate_args = {
+            arg: hook_args.annotations[arg]
+            for arg in hook_args.annotations
+            if arg != 'return'
+        }
+
+        if len(annotate_args) != len(hook_args.args):
+            raise TypeError("All hooks arguments must have the type informed")
+
+        if not inspect.getdoc(hook):
+            raise TypeError("All hooks must have documentation")
+
+
+class HookMan():
+    """
+    Main class of HookMan, this class holds all the information related to the plugins
+    """
+
+    def __init__(self, *, specs: HooksSpecs, plugin_dirs: List[Path]):
+        self.specs = specs
+
+        config_files = []
+        for plugin_dir in plugin_dirs:
+            config_files += plugin_dir.glob('**/plugin.yaml')
+
+        if config_files:
+            self.plugin_config_files = config_files
+        else:
+            raise FileNotFoundError("The given path doesn't have a .yaml file")
+
+        self.hooks_available = {
+            f'{hook.__name__.lower()}': f'{specs.project_name}_v{specs.version}_{hook.__name__.lower()}'
+            for hook in specs.hooks
+        }
+
+    def get_hook_caller(self) -> 'HookCaller':
+        """
+        Return a HookCaller class that holds all references for the functions implemented on the plugins.
+        """
+        # self._spec.pyd_name = '_alfasim_hookcaller'
+        _hookman = __import__(self.specs.pyd_name)
+        hook_caller = _hookman.HookCaller()
+
+        dll_path_locations = self._get_dll_locations()
+        for dll_path in dll_path_locations:
+            self._bind_dll_functions_on_hook_caller(dll_path, hook_caller)
+
+        return hook_caller
+
+    def _get_dll_locations(self) -> List[Path]:
+        """
+        Load the config file of each plugin available and return a list with the location of all DLL.
+        """
+        schema = strictyaml.Map({
+            "author": strictyaml.Str(),
+            "contact_information": strictyaml.Str(),
+            "plugin_version": strictyaml.Str(),
+            "dll_name": strictyaml.Str()
+        })
+
+        dll_locations = []
+        for hook_config_file in self.plugin_config_files:
+            plugin_config_content = strictyaml.load(hook_config_file.read_text(), schema).data
+            dll_name = plugin_config_content['dll_name']
+            dll_locations.append(hook_config_file.parent / dll_name)
+
+        return dll_locations
+
+    def _bind_dll_functions_on_hook_caller(self, dll_path: Path, hook_caller : 'HookCaller'):
+        """
+        Load the dll_path from the plugin and bind methods that are implemented on the hook_caller
+        """
+        plugin_dll = ctypes.cdll.LoadLibrary(str(dll_path))
+
+        hooks_to_bind = {

```

Comment:
So this is binding the *last* plugin which implements a certain function, is that what you intended? I remember we deciding it should raise an error if more than one plugin implemented a hook function.

Reply:
This logic of "if more than one plugin implemented a hook function" I was planning to do with the next task (to list all implemented hooks and so on).